### PR TITLE
`Axes3D.plot_surface`: Allow masked arrays and `NaN` values

### LINKED ISSUE #487: Plotting masked arrays with plot_surface()
I would be nice if `plot_surface()` could plot masked arrays. Unlike `imshow()` which really plots the masked data,
`plot_surface()` only plots the non-masked data.
Here is an example:

```
import numpy as np
import pylab as mpl
from mpl_toolkits.mplot3d import axes3d

x = np.arange(1,10,1)
y = np.arange(1,10,1)
x,y = np.meshgrid(x,y)
z = x**3 + y**3 - 500
z = np.ma.masked_array(z, z<0)

cm = mpl.cm.jet

ax1 = mpl.subplot(1,2,1, projection='3d')
ax1.plot_surface(x,y,z,
                 rstride=1, cstride=1, linewidth=0,
                 cmap=cm)

ax2 = mpl.subplot(1,2,2)
ax2.imshow(z, cmap=cm)

mpl.show()
```

### LINKED ISSUE #4941: Axes3d plot_surface not supporting masked arrays?
Consider this code:

```
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.colors as colors
from matplotlib import cm
from mpl_toolkits.mplot3d import Axes3D

fig = plt.figure()
ax = fig.add_subplot(1, 1, 1, projection='3d')

x, y = np.meshgrid(np.linspace(-1, 1), np.linspace(-1, 1))
z = x**2 + y**2

z[10:20, 10:20] = 0.0

masked_array = np.ma.masked_equal(z, 0.0)
palette = cm.gray
palette.set_over('g', 1.0)
palette.set_under('b', 1.0)
palette.set_bad('r', 1.0)

ax.plot_surface(x,y,masked_array,
    rstride=1,
    cstride=1,
    cmap=palette,
    vmin=-1,
    vmax=1,
    )
plt.show()
```

The area set to zero should be red, but it isn't. It works on a 2d axis using pcolormesh.

### LINKED ISSUE #8222: matplotlib 3D surface - gaps / holes in surface
### Bug report

**Bug summary**

A grid is interpolated with scipy's `griddata()` and contains values and NaNs. However, when the grid is plotted with mplot3d's `plot_surface()` there are gaps in the surface (see upper plot). The lower the view height, the more of such gaps are rendered. When the grid is plotted in 2D with `imshow()` there are no gaps (see lower plot).

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

Z = np.loadtxt('data.txt')
X = np.linspace(0,3,301)
Y = np.linspace(0,3,301)
X = np.array([X,]*(301)).transpose()
Y = np.array([Y,]*(301))
fig = plt.figure(figsize=(15,15), dpi=100)
ax = fig.add_subplot(2, 1, 1, projection='3d')
ax.view_init(40,300)
surf = ax.plot_surface(X, Y, Z, cmap='jet', rstride=1, cstride=1, antialiased=False, shade=False, alpha=1.0, linewidth=0, vmin=0.25, vmax=0.35);

ax.invert_yaxis()
ax.dist = 11

ax = fig.add_subplot(2, 1, 2)
plt.imshow(Z, cmap='jet', vmin=0.25, vmax=0.35, origin='lower');

plt.show()

```

**Actual outcome**

Close-up:

**Expected outcome**

No holes in the 3D surface plot.

**Matplotlib version**

Matplotlib 1.3.0 to 1.5.3 (WinPython 3.3.2.3 to 3.5.2.3), but it was also reproduced up to Matplotlib 2.0: [http://stackoverflow.com/questions/41268550/matplotlib-3d-surface-gaps-holes-in-surface](http://stackoverflow.com/questions/41268550/matplotlib-3d-surface-gaps-holes-in-surface)

### LINKED ISSUE #12395: Rendering issue occurs when plotting 3D surfaces at a discontinuity
### Bug report

**Bug summary**

When running the code below to plot the surface for the equation 1 - y/x, the plot works but I get an error and the rendering of the surface breaks when viewed from some angles.

**Code for reproduction**

```python
import math
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.ticker
from matplotlib import cm
from mpl_toolkits.mplot3d import Axes3D

# Generate the data
x = np.linspace(-1.0, 1.0, 50)
y = np.linspace(-1.0, 1.0, 50)
z = np.zeros(len(x))
x,y = np.meshgrid(x,y)
z = 1 - y/x
z[z >= 5.0] = 5.0            # Clips top of surface
z[z <= -5.0] = -5.0          # Clips bottom of surface

# place NaNs at the discontinuity
pos = np.where(np.abs(np.diff(z)) >= 5.0)
z[pos] = np.nan

# Create the plot
p1 = plt.figure()
ax = p1.gca(projection='3d')
surf = ax.plot_surface(x,y,z, rstride=1, cstride=1, cmap=cm.coolwarm, linewidth=0,
                           vmin=np.nanmin(z), vmax=np.nanmax(z), antialiased=False)
plt.title("1 - y/x")
ax.set_xlim(-1.0, 1.0)
ax.set_ylim(-1.0, 1.0)
ax.set_zlim(-5.0, 5.0)
ax.set_xlabel("x")
ax.set_ylabel("y")
ax.set_zlabel("z")
p1.show()
raw_input("Press any key to close all plots: ")
```

**Actual outcome**

```
/usr/lib/pymodules/python2.7/matplotlib/colors.py:576: RuntimeWarning: invalid value encountered in less
  cbook._putmask(xa, xa < 0.0, -1)
```

Odd rendering issues appear:

rendering_angle1

render_angle3

**Expected outcome**

I'd expect to not get errors and to not see the rendering issue. I am using older versions of virtually everything, and this may be a platform-specific issue. Or I may be using this code in a way it was not intended to be used, in which case if there is a more-correct way to do this I'd appreciate someone showing me how to do it.

**Matplotlib version**

  * Operating system: Ubuntu 14.04
  * Matplotlib version: 1.3.1
  * Matplotlib backend (`print(matplotlib.get_backend())`): 'TkAgg'
  * Python version: 2.7.6
  * Jupyter version (if applicable): N/A
  * Other libraries:

### LINKED ISSUE #16470: Inconsistent Corner Masking w/ plot_surface
### Inconsistent Corner Masking w/ plot_surface

plot_surface inconsistently corner masks and even may not plot certain values when given a non-rectangular array.

```python

import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl
from mpl_toolkits.mplot3d import axes3d

x = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
y = [1, 2, 3, 4, 5, 6, 7, 8]

X, Y = np.meshgrid(x, y)

matrix = np.array([[np.nan, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
                    [np.nan, 1.0, 2.0, 3.0, 4.0, 4.0, 4.0, 3.0, 2.0, 1.0, 1.0],
                    [np.nan, np.nan, 4.0, 5.0, 6.0, 8.0, 6.0, 5.0, 4.0, 3.0, np.nan],
                    [np.nan, np.nan, 7.0, 8.0, 11.0, 12.0, 11.0, 8.0, 7.0, np.nan, np.nan],
                    [np.nan, np.nan, 8.0, 9.0, 10.0, 16.0, 10.0, 9.0, 10.0, 7.0, np.nan],
                    [np.nan, np.nan, np.nan, 12.0, 16.0, 20.0, 16.0, 12.0, 11.0, np.nan, np.nan],
                    [np.nan, np.nan, np.nan, np.nan, 22.0, 24.0, 22.0, 20.0, 18.0, np.nan, np.nan],
                    [np.nan, np.nan, np.nan, np.nan, np.nan, 28.0, 26.0, 25.0, np.nan, np.nan, np.nan]])
maskMat = np.ma.masked_invalid(matrix)
print(maskMat)

fig = plt.figure()
ax = fig.gca(projection = '3d')
normC = mpl.colors.Normalize(vmax = maskMat.max(), vmin = maskMat.min())
colors = mpl.cm.plasma(normC(maskMat))
surf = ax.plot_surface(X, Y, maskMat, norm = normC)
plt.show()
```

**Actual outcome**
Corner masking only happens on the left and not on the right:
bug report

A case where a point is dropped:
bug report II

**Expected outcome**
Corner masking consistently, and/or a way to control it like in contourf

**Matplotlib version**

  * Operating system: Windows 10
  * Matplotlib version: 3.1.3
  * Matplotlib backend (`print(matplotlib.get_backend())`): TkAgg
  * Python version: 3.7.3
  * Other libraries: Numpy 1.18.1

Matplotlib was installed via PyCharm Community edition

### PR DESCRIPTION
**Pull request by tomneep**
## PR Summary

- Closes #487
- Closes #4941 
- Closes #8222 
- Closes #12395
- Closes #16470

This is my third attempt at solving the issues with using masked arrays in `Axes3D.plot_surface`, previous attempts are #18114 and #20646. Hopefully it is third time lucky.

I got the impression that the feedback on #18114 was reasonably positive but this approach solves more issues and also behaves like `contour` with `corner_mask=True` which was a preference expressed by @ianthomas23.

The suggestion from #20646 was that it would be better to deal with NaNs before we get to `Poly3DCollection.do_3d_projection` which is called every time we adjust an interactive figure.

I've gone "all in" on this PR and removed the warning about not supporting NaNs and added masked array support, if you prefer to take this step-by-step e.g. removing the masked array part or keeping the NaN warning, then let me know, and I'll happily change.

There are no tests yet, once I've had some feedback on this I'll add some based on the issues that this closes.

Examples of the issues that this PR solves (along with the code to make them) are below. I didn't add the example from #8222 which loads a large dataset but the before and after are visually the same as the examples shown in #20646.

### Examples before this PR

plot_surface_master

### Examples after this PR 

plot_surface_fixes

### Code for examples 
```python
import matplotlib
import matplotlib.pyplot as plt
import numpy as np

fig, axes = plt.subplots(
    nrows=2,
    ncols=2,
    figsize=(10, 10),
    subplot_kw={"projection": "3d"},
    constrained_layout=True,
)

# 487
x, y = np.mgrid[1:10:1, 1:10:1]
z = x ** 3 + y ** 3 - 500
z = np.ma.masked_array(z, z < 0)

plot_opts = dict(rstride=1, cstride=1, linewidth=0, cmap="inferno")
axes[0, 0].plot_surface(x, y, z, **plot_opts)
axes[0, 0].view_init(35, -90)
axes[0, 0].set(title="Issue #487")

# 4941
x, y = np.mgrid[-1:1:0.05, -1:1:0.05]
z = (x ** 2 + y ** 2) + 0.01
z[10:20, 10:20] = 0.0

masked_array = np.ma.masked_equal(z, 0.0)
palette = plt.get_cmap("gray").copy()
palette.set_over("g", 1.0)
palette.set_under("b", 1.0)
palette.set_bad("r", 1.0)

plot_opts = dict(rstride=1, cstride=1, cmap=palette, vmin=-1, vmax=1)
axes[0, 1].plot_surface(x, y, masked_array, **plot_opts)
axes[0, 1].view_init(35, -90)
axes[0, 1].set(title="Issue #4941")

# 12395
z = (1 - y / x).clip(min=-5, max=5)
# place NaNs at the discontinuity
pos = np.where(np.abs(np.diff(z, axis=0)) >= 5.0)
z[pos] = np.nan

plot_opts = dict(rstride=1, cstride=1, cmap="coolwarm", antialiased=False)
axes[1, 0].plot_surface(x, y, z, **plot_opts)
axes[1, 0].set(title="Issue #12395")

#16470
x = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
y = [1, 2, 3, 4, 5, 6, 7, 8]

x, y = np.meshgrid(x, y)
nan = np.nan
matrix = np.array(
    [
        [nan, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
        [nan, 1.0, 2.0, 3.0, 4.0, 4.0, 4.0, 3.0, 2.0, 1.0, 1.0],
        [nan, nan, 4.0, 5.0, 6.0, 8.0, 6.0, 5.0, 4.0, 3.0, nan],
        [nan, nan, 7.0, 8.0, 11.0, 12.0, 11.0, 8.0, 7.0, nan, nan],
        [nan, nan, 8.0, 9.0, 10.0, 16.0, 10.0, 9.0, 10.0, 7.0, nan],
        [nan, nan, nan, 12.0, 16.0, 20.0, 16.0, 12.0, 11.0, nan, nan],
        [nan, nan, nan, nan, 22.0, 24.0, 22.0, 20.0, 18.0, nan, nan],
        [nan, nan, nan, nan, nan, 28.0, 26.0, 25.0, nan, nan, nan],
    ]
)
z = np.ma.masked_invalid(matrix)
normC = matplotlib.colors.Normalize(vmax=z.max(), vmin=z.min())
colors = plt.get_cmap("plasma")(normC(z))
axes[1, 1].plot_surface(x, y, z, facecolors=colors)
axes[1, 1].view_init(30, -80)
axes[1, 1].set(title="Issue #16470")
```

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 1]
```diff
@@ -1655,6 +1651,24 @@ def plot_surface(self, X, Y, Z, *args, norm=None, vmin=None,
...
                         colset.append(fcolors[rs][cs])
 
+        # In cases where there are NaNs in the data (possibly from masked
+        # arrays), artifacts can be introduced. Here check whether NaNs exist
+        # and remove the entries if so
+        if not isinstance(polys, np.ndarray) or np.isnan(polys).any():
+            new_polys = []
+            new_colset = []
+
+            for p, col in itertools.zip_longest(polys, colset):
```
Comment (timhoffm) [inline lib/mpl_toolkits/mplot3d/axes3d.py:1661]: This is quite clever for handling the two `fcolor` cases. However, I think we should explain it in a comment.

> depending on fcolors, colset is either an empty list or has as many elements as polys.
> In the former case new_colset results in a list with None entries, that is discarded later.

**Review by timhoffm — COMMENTED** [review]
I'm slightly worried that we are throwing away data and not storing the original inputs in the `PolyCollection`. Do we have precedent for that?

Nevertheless I'm tempted to accept the solution as it is much simpler and faster than any later data cleaning. Maybe we should add a note on the data cleaning in the docstring.

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 1 above)
Comment (tomneep) [reply lib/mpl_toolkits/mplot3d/axes3d.py:1661]: I've added your suggested comment locally, I'll wait for further changes (e.g. adding tests) before pushing to save running the CI

**Comment by tomneep** [comment]
> I'm slightly worried that we are throwing away data and not storing the original inputs in the `PolyCollection`. Do we have precedent for that?

I don't know, hopefully someone else does. The one thought I had was to see what is done in `contour/contourf`. Here it looks like the masked data are not saved in the `collections` (but perhaps I am missing something)? Would this be an equivalent case?

```python
import matplotlib.pyplot as plt
import numpy as np

x, y = np.mgrid[-10:10.1:2, -10:10.1:2]
z = np.ma.masked_less(x ** 2 + y ** 2, 50)

fig, ax = plt.subplots(figsize=(6, 6))
contours = ax.contourf(x, y, z)

segs = np.concatenate([np.concatenate(s) for s in contours.allsegs if s]).T
paths = np.concatenate(
    [i.vertices for paths in contours.collections for i in paths.get_paths()]
).T

ax.plot(segs[0], segs[1], "rs")
ax.plot(paths[0], paths[1], ".b")
```
gives
contour_test

**Comment by jklymak** [comment]
The user threw the data away when they NaN-ed or masked it.  I don't think we are going to offer a way to unmask it after the surface is made, so I don't think there is harm in removing it.  But maybe I'm missing a subtlety in the data->polygon pipeline.  I assume if the user changes x, y, Z in the data then the polygons are regenerated anyway?

**Comment by tomneep** [comment]
Sorry @jklymak I don't understand your question. What kind of changing of x, y and Z do you have in mind?

**Comment by jklymak** [comment]
via `set_data` or whatever the 3-d equivalent is...

**Comment by tomneep** [comment]
I'm not sure what the equivalent would be. You could get the paths with `get_paths` and perhaps modify those, but it would be tricky. Maybe there is some easy method that I am missing. 

The original x, y and z don't get passed to the `Poly3DCollection`. I suppose I could imagine a new collection that computed the polygons from the x, y and z and would recalculate the polygons if `set_data` was called, but that isn't how it currently works. Is that what you had in mind?

**Comment by jklymak** [comment]
Sorry if I'm taking you off on a tangent - many of our methods save a representation of the original data.  If this has no such representation, then there really is no harm done in throwing out masked data because the user can't modify it post-facto anyways.

**Comment by tomneep** [comment]
The `Poly3DCollection` is made directly from the vertices of the polygons and doesn't see the original data. It isn't just the masking because there are the `stride/count` options in `plot_surface` that (can) also throw data away before the polygons are made. So I *think* we are ok on that front?

**Comment by jklymak** [comment]
@eric-wieser @WeatherGod @ianthomas23 do any of you have time to comment on _this_ solution to the masking problem?   It seems reasonable to _me_ but I don't know if I've ever used `plot_surface` in matplotlib ;-)  Thanks!

**Review by timhoffm — APPROVED** [review]

**Comment by eric-wieser** [comment]
Can you add a test showing how this behaves with striding?

**Comment by tomneep** [comment]
@eric-wieser Here is the example you request (for before and after this PR). It is a bit hard to show here but from the side angle (third plot) it doesn't look right, neither before and after this PR. Is this what you wanted to see?

#### Example without PR

stride_tests_master


#### Example with PR

stride_tests_pr

#### Code for plots
<details>

```python
import matplotlib.pyplot as plt
import numpy as np

x, y = np.mgrid[-2:2.1:1, -2:2.1:1]
z = np.ma.masked_less(x * y, 2)

fig, (ax1, ax2, ax3) = plt.subplots(
    figsize=(9, 3), ncols=3, subplot_kw={"projection": "3d"}
)
ax1.plot_surface(x, y, z)
ax1.set_title("Default strides")
ax1.view_init(60, -45)

surf = ax2.plot_surface(x, y, z, rstride=4, cstride=4)
ax2.set_title("r/cstride = 4")
ax2.view_init(60, -45)

surf = ax3.plot_surface(x, y, z, rstride=4, cstride=4)
ax3.set_title("r/cstride = 4")
ax3.view_init(2, 30)
```

</details>

**Comment by eric-wieser** [comment]
Thanks, that does show what I was expecting to see. I don't think the side view is interesting, but I think the middle plot might be worth adding a test for in case someone decides they want different behavior later. Using `np.mgrid[-6:6.1:1, -6:6.1:1]` but leaving the stride at 4 would probably produce a slightly more insightful plot for deciding whether the current behavior is sensible.

**Comment by tomneep** [comment]
Thanks, I've added a test with the range you suggest. Indeed the plot looks a bit more interesting than my previous comment, with a double arrowhead/bowtie arrangement. Does this look ok to you?

strides test

**Comment by eric-wieser** [comment]
Yes, that looks like a good test case to include, thanks! My question (possibly for a future PR) would be whether it makes sense to split that middle polygon in two, into two triangles - but it's not immediately clear to me how such an approach would generalize.

**Comment by tomneep** [comment]
Yes I had the same thought after the first example (whether the polygon could be split into pieces to help the side view). I'm not sure how simple it is, so I think as you say it might be a question for a future PR.

**Comment by jklymak** [comment]
I'm not sure we should be worried about striding effects - if you decimate your data you should expect aliasing, and if you decimate a mask it is luck of the draw where on the mask your stride will fall.  If you need more control, it really isn't hard to stride your own data.

**Comment by eric-wieser** [comment]
I'd dispute that the claim that it's easy to stride your own data - this isn't just simple striding that takes 1/n² of the elements, it's preserving full resolution along the gridlines and discarding data in the cells, taking (2n-1)/n² of the elements. That means there's no way to represent the striding in a numpy array.

That's not to say you're wrong in saying that the weirdness in the plot above is unimportant - I think it's fine to merge with the current behavior, and was only pushing for a test to ensure we were aware of the quirk.

**Comment by jklymak** [comment]
> this isn't just simple striding that takes 1/n² of the elements, it's preserving full resolution along the gridlines and discarding data in the cells, taking (2n-1)/n² of the elements. That means there's no way to represent the striding in a numpy array.

probably off topic, but that is confusing.  The docs say `rstride, cstride : Downsampling stride in each direction`.  Given that, I'm not sure what "full resolution along the gridlines and discarding data in the cells" means.

**Review by ianthomas23 — APPROVED** [review]
There were a couple of CI failures, but I've rerun them and they now pass.

As my opinion was requested, I've taken a look at this PR and am happy with the output.  But I am not a frequent user of mplot3d.

**Comment by eric-wieser** [comment]
> > this isn't just simple striding that takes 1/n² of the elements, it's preserving full resolution along the gridlines and discarding data in the cells, taking (2n-1)/n² of the elements. That means there's no way to represent the striding in a numpy array.
> 
> probably off topic, but that is confusing. The docs say `rstride, cstride : Downsampling stride in each direction`. Given that, I'm not sure what "full resolution along the gridlines and discarding data in the cells" means.

I mean that stride=3  samples as
```
#######
#  #  #
#  #  #
#######
#  #  #
#  #  #
#######
```
Not
```
#  #  #


#  #  #


#  #  #
```

**Comment by eric-wieser** [comment]
The docs are using "stride" in a vague sense, so they're not wrong per se, but certainly easy to misread. The reading I have of them that is consistent with the implementation is "Split the surface into row and column gridlines spaced `rstride` and `cstride` entries apart, but draw the perimeters of the resulting grid squares without downsampling". I suppose this phrasing in the docs is actively misleading:

> If the input data is larger, it will be downsampled (by slicing) to these numbers of points.

I think the current behavior of the striding (ignoring this PR) is the right behavior, but the docs don't describe it well.

**Comment by QuLogic** [comment]
It appears this is good to go, and the striding discussion is orthogonal, so I'm going to merge this.

**Pull request merged**